README.md
# gkchestertonr <img src="tools/logo.png" align="right"/>

[![Build Status](https://travis-ci.org/bigpas/gkchestertonr.svg?branch=master)](https://travis-ci.org/bigpas/gkchestertonr)

## An R Package for a selection of G.K. Chesterton's books

> "A good novel tells us the truth about its hero;
> but a bad novel tells us the truth about its author."

(*Heretics*)

This package provides access to the full texts of a selection of G.K. Chesterton's published books. The UTF-8 plain text for each novel was sourced from [Project Gutenberg](https://www.gutenberg.org/), processed a bit, and is ready for text analysis. Each text is in a character vector with elements of about 70 characters. The package contains:

There is also a function `chesterton_books()` that returns a tidy data frame of all selected novels.

## Installation

You can install the development version from Github:

```
library(devtools)
install_github("bigpas/gkchestertonr")
library(gkchestertonr)
```

NEWS.md
# gkchestertonr 0.1.0

* Initial release of full texts of G.K. Chesterton's complete works.


tests/testthat.R
library(testthat)
library(gkchestertonr)

test_check("gkchestertonr")


tests/testthat/test_chesterton_books.R
# tests for chesterton_books function

context("Tidy dataframe for books")

suppressPackageStartupMessages(library(dplyr))

test_that("factor order is correct", {
        d <- chesterton_books()
        expect_equal(levels(d$book)[1], "Heretics")
        expect_equal(levels(d$book)[2], "Orthodoxy")
})

test_that("tidy frame for Chesterton books is right", {
        
        d <- chesterton_books() %>%
                group_by(book) %>%
                summarise(total_lines = n())
        
        expect_equal(nrow(d), 11)
        expect_equal(ncol(d), 2)
        expect_equal(as.character(d$book[1]), "Heretics")
        expect_equal(as.character(d$book[3]), "Man Alive")
        expect_lt(d$total_lines[1], d$total_lines[3])
})


R/novels.R
#' G.K. Chesterton's Complete Works.
#' 
#' This package contains the complete text of G.K. Chesterton's works, 
#' formatted to be convenient for text analysis.
#' @docType package
#' @name gkchestertonr
#' @aliases gkchestertonr gkchestertonr-package
NULL

#' The text of G.K. Chesterton's "Heretics"
#'
#' A dataset containing the text of Heretics.
#' It is a collection of 20 essays originally published by 
#' G.K. Chesterton in 1905. 
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/470}
#' @format A character vector with xxx elements
"heretics"

#' The text of G.K. Chesterton's "Orthodoxy"
#'
#' A dataset containing the text of G.K. Chesterton's 1908, 
#' a classic of Christian apologetics. 
#' The book was written when Chesterton was an Anglican. 
#' 14 years later he converted to Catholicism. 
#' Chesterton chose the title, Orthodoxy, 
#' to focus instead on the plainness of the Apostles' Creed.
#' 
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/130}
#' @format A character vector with xxx elements
"orthodoxy"

#' The text of G.K. Chesterton's novel "Man Alive"
#'
#' A dataset containing the text of G.K. Chesterton's 1912 novel "Man Alive". 
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/1718}
#' @format A character vector with xxx elements
"manalive"

#' The text of G.K. Chesterton's novel ""
#'
#' A dataset containing the text of G.K. Chesterton's 1815 novel "Emma".
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each. 
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/158}
#' @format A character vector with xxx elements
"thenewjerusalem"

#' The text of G.K. Chesterton's novel The Man Who Was Thursday: A Nightmare
#'
#' A dataset containing the text of G.K. Chesterton's novel 
#' "The Man Who Was Thursday: A Nightmare", 
#' first published in 1908
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/1695}
#' @format A character vector with xxx elements
"themanwhowasthursday"

#' The text of G.K. Chesterton's novel "What's Wrong with the World"
#'
#' A dataset containing the text of G.K. Chesterton's collection of essays 
#' "What's Wrong with the World", published in 1910.
#' A collection of classic essays by the inimitable G.K. Chesterton, 
#' exploring a variety of topics.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/1717}
#' @format A character vector with xxx elements
"whatswrongwiththeworld"

#' The text of G.K. Chesterton's novel "The Napoleon of Nothing Hill"
#'
#' A dataset containing the text of G.K. Chesterton's novel 
#' "The Napoleon of Nothing Hill", published in 1904.

#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/20058}
#' @format A character vector with xxx elements
"thenapoleonofnothinghill"

#' The text of G.K. Chesterton's "Eugenics and other Evils"
#'
#' A dataset containing the text of G.K. Chesterton's "Eugenics and other Evils", 
#' published in 1922.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/25308}
#' @format A character vector with xxx elements
"eugenicsandotherevils"

#' The text of G.K. Chesterton's "A Short History of England"
#'
#' A dataset containing the text of G.K. Chesterton's 
#' "A Short History of England", 
#' published in 1917.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/20897}
#' @format A character vector with xxx elements
"ashorthistoryofengland"

#' The text of G.K. Chesterton's "The Man Who Knew too much"
#'
#' A dataset containing the text of G.K. Chesterton's detective story 
#' "The Man Who Knew too much", 
#' published in 1922.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/1720}
#' @format A character vector with xxx elements
"themanwhoknewtoomuch"

#' The text of G.K. Chesterton's "The Ballad of the White Horse"
#'
#' A dataset containing the text of G.K. Chesterton's  
#' "The Ballad of the White Horse", published in 1911.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/1719}
#' @format A character vector with xxx elements
"theballadofthewhitehorse"


R/chesterton_books.R
#' Tidy data frame of G.K. Chesterton's 6 works.
#' 
#' Returns a tidy data frame of G.K. Chesterton's completed and published works with 
#' two columns: \code{text}, which contains the text of the novels divided into 
#' elements of up to about 70 characters each, and \code{book}, which contains the titles of
#' the novels as a factor in order of publication.
#' 
#' @details Users should be aware that there are some differences in usage 
#' between the novels as made available by Project Gutenberg. For example, 
#' "anything" vs. "any thing", "Mr" vs. "Mr.", and using underscores vs. all 
#' caps to indicate italics/emphasis.
#' 
#' @return A data frame with two columns: \code{text} and \code{book}
#' 
#' @name chesterton_books
#' 
#' @examples 
#' 
#' library(dplyr)
#'
#' chesterton_books() %>% group_by(book) %>%
#'      summarise(total_lines = n())
#'
#' @export
chesterton_books <- function(){
        books <- list(
                "Heretics" = gkchestertonr::heretics,
                "Orthodoxy" = gkchestertonr::orthodoxy,
                "Man Alive" = gkchestertonr::manalive,
                "The Man Who Was Thursday" = gkchestertonr::themanwhowasthursday,
                "The Man Who Knew Too Much" = gkchestertonr::themanwhoknewtoomuch,
                "What's Wrong with the World" = gkchestertonr::whatswrongwiththeworld,
                "The New Jerusalem" = gkchestertonr::thenewjerusalem,
                "What's Wrong with the World" = gkchestertonr::whatswrongwiththeworld,
                "The Napoleon of Nothing Hill" = gkchestertonr::thenapoleonofnothinghill,
                "Eugenics and other Evils" = gkchestertonr::eugenicsandotherevils,
                "The Ballad of the White Horse" = gkchestertonr::theballadofthewhitehorse,
                "A Short History of England" = gkchestertonr::ashorthistoryofengland
        )
       ret <- data.frame(text = unlist(books, use.names = FALSE), 
                         stringsAsFactors = FALSE)
       ret$book <- factor(rep(names(books), sapply(books, length)))
       ret$book <- factor(ret$book, levels = unique(ret$book))
       structure(ret, class = c("tbl_df", "tbl", "data.frame"))
}


data-raw/prep_data.R
## This is the script used to download the Project Gutenberg text files 

library(gutenbergr)

get_book <- function(gutenber_ebook_num, title_line, b1=NULL, b2=NULL, tail_blank = 0) {
        
        book <- gutenberg_download(gutenber_ebook_num)[['text']]
        if(!is.null(b1) & !is.null(b2)){
                return(c(book[title_line:b1], book[b2:length(book) - tail_blank]))                
        } else {
                return(book[title_line:length(book) - tail_blank])
        }
        
}


orthodoxy            <- get_book(gutenber_ebook_num = 130, title_line = 6)
themanwhowasthursday <- get_book(gutenber_ebook_num = 1695, title_line = 1)
heretics             <- get_book(gutenber_ebook_num = 470, title_line = 1, b1 = 11, b2 = 84)
themanwhoknewtoomuch <- get_book(gutenber_ebook_num = 1720, title_line = 8)

thenapoleonofnothinghill <- get_book(gutenber_ebook_num = 20058, title_line = 13, b1=20, b2=39)
manalive                 <- get_book(gutenber_ebook_num = 1718, title_line = 10, b1=12, b2=48)
eugenicsandotherevils    <- get_book(25308, title_line = 19, b1=20, b2=39)
thenewjerusalem          <- get_book(13468, 1, b1=6, b2=12)

whatswrongwiththeworld   <- get_book(gutenber_ebook_num = 1717, title_line = 1)
theballadofthewhitehorse <- get_book(gutenber_ebook_num = 1719, title_line = 1)
ashorthistoryofengland   <- get_book(gutenber_ebook_num = 20897, title_line = 1, b1=6, b2=24)

devtools::use_data(themanwhowasthursday, overwrite = TRUE)
devtools::use_data(orthodoxy, overwrite = TRUE)
devtools::use_data(whatswrongwiththeworld, overwrite = TRUE)
devtools::use_data(theballadofthewhitehorse, overwrite = TRUE)
devtools::use_data(heretics, overwrite = TRUE)
devtools::use_data(ashorthistoryofengland, overwrite = TRUE)
devtools::use_data(thenapoleonofnothinghill, overwrite = TRUE)
devtools::use_data(manalive, overwrite = TRUE)
devtools::use_data(eugenicsandotherevils, overwrite = TRUE)
devtools::use_data(thenewjerusalem, overwrite = TRUE)
devtools::use_data(themanwhoknewtoomuch, overwrite = TRUE)

## Finished!

## Making a hex sticker for gkchestertonr

hexSticker::sticker("https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/GK_Chesterson_signature.svg/585px-GK_Chesterson_signature.svg.png", 
                    package="", 
                    spotlight = FALSE,
                    h_size = 1.5,
                    h_color = "black",
                    h_fill = "#009933",
                    p_size=0.5, p_y = 0.1,
                    p_family = "RobotoCondensed-Regular",
                    s_x=1, 
                    s_y=0.95, 
                    s_width=0.96,
                    url="gkchestertonr",
                    u_size = 2,
                    u_y = 0.1,
                    u_color = "black",
                    filename="tools/gkchestertonr.png")

gk <- magick::image_read("tools/gkchestertonr.png")
image_scale(gk, "150") %>%
        image_write(path = "tools/logo.png", format = "png")



R/books.R
#' G.K. Chesterton's Complete Works.
#' 
#' This package contains the complete text of G.K. Chesterton's works, 
#' formatted to be convenient for text analysis.
#' @docType package
#' @name gkchestertonr
#' @aliases gkchestertonr gkchestertonr-package
NULL

#' The text of G.K. Chesterton's "Heretics"
#'
#' A dataset containing the text of Heretics.
#' It is a collection of 20 essays originally published by 
#' G.K. Chesterton in 1905. 
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/470}
#' @format A character vector with 5630 elements
"heretics"

#' The text of G.K. Chesterton's "Orthodoxy"
#'
#' A dataset containing the text of G.K. Chesterton's 1908, 
#' a classic of Christian apologetics. 
#' The book was written when Chesterton was an Anglican. 
#' 14 years later he converted to Catholicism. 
#' Chesterton chose the title, Orthodoxy, 
#' to focus instead on the plainness of the Apostles' Creed.
#' 
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/130}
#' @format A character vector with 5839 elements
"orthodoxy"

#' The text of G.K. Chesterton's novel "Man Alive"
#'
#' A dataset containing the text of G.K. Chesterton's 1912 novel "Man Alive". 
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/1718}
#' @format A character vector with 6336 elements
"manalive"

#' The text of G.K. Chesterton's novel "The New Jerusalem"
#'
#' A dataset containing the text of G.K. Chesterton's "The New Jerusalem".
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each. 
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/158}
#' @format A character vector with 8377 elements
"thenewjerusalem"

#' The text of G.K. Chesterton's novel The Man Who Was Thursday: A Nightmare
#'
#' A dataset containing the text of G.K. Chesterton's novel 
#' "The Man Who Was Thursday: A Nightmare", 
#' first published in 1908
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/1695}
#' @format A character vector with 6462 elements
"themanwhowasthursday"

#' The text of G.K. Chesterton's novel "What's Wrong with the World"
#'
#' A dataset containing the text of G.K. Chesterton's collection of essays 
#' "What's Wrong with the World", published in 1910.
#' A collection of classic essays by the inimitable G.K. Chesterton, 
#' exploring a variety of topics.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/1717}
#' @format A character vector with 11076 elements
"whatswrongwiththeworld"

#' The text of G.K. Chesterton's novel "The Napoleon of Nothing Hill"
#'
#' A dataset containing the text of G.K. Chesterton's novel 
#' "The Napoleon of Nothing Hill", published in 1904.

#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/20058}
#' @format A character vector with 6409 elements
"thenapoleonofnothinghill"

#' The text of G.K. Chesterton's "Eugenics and other Evils"
#'
#' A dataset containing the text of G.K. Chesterton's "Eugenics and other Evils", 
#' published in 1922.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/25308}
#' @format A character vector with 4817 elements
"eugenicsandotherevils"

#' The text of G.K. Chesterton's "A Short History of England"
#'
#' A dataset containing the text of G.K. Chesterton's 
#' "A Short History of England", 
#' published in 1917.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/20897}
#' @format A character vector with 5367 elements
"ashorthistoryofengland"

#' The text of G.K. Chesterton's "The Man Who Knew too much"
#'
#' A dataset containing the text of G.K. Chesterton's detective story 
#' "The Man Who Knew too much", 
#' published in 1922.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/1720}
#' @format A character vector with 6449 elements
"themanwhoknewtoomuch"

#' The text of G.K. Chesterton's "The Ballad of the White Horse"
#'
#' A dataset containing the text of G.K. Chesterton's  
#' "The Ballad of the White Horse", published in 1911.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/1719}
#' @format A character vector with 3341 elements
"theballadofthewhitehorse"
